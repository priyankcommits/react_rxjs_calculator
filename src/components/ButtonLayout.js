import Button from 'react-bootstrap/Button';
import Col from 'react-bootstrap/Col';
import React from 'react';
import Row from 'react-bootstrap/Row';

import calculatorButtons from '../constants';
import numberEvent$ from '../events';

function ButtonLayout() {
  
  const publishButtonEvent = (value, type) => {
    numberEvent$.next({value, type});
  }

  let chunkedCalculatorButtons = [];
  for(let i = 0; i < calculatorButtons.length; i += 6)
    chunkedCalculatorButtons.push(calculatorButtons.slice(i, i + 6));

  console.log('Rendering Button Layout');
  return (
    <div>
      {chunkedCalculatorButtons.map((chunk, index) => (
        <Row key={index} className='mt-2'>
          {chunk.map((button, buttonIndex) => (
            <Col key={buttonIndex} xs={2}>
              <Button className='w-100' variant={button.color} onClick={() => publishButtonEvent(button.value, button.type)}>
                {button.value}
              </Button>
            </Col>
          ))}
        </Row>
      ))}
    </div>
  );
}

export default ButtonLayout;
