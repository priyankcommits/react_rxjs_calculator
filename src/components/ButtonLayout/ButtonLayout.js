// Component to display all claculator buttons from constants file
// Publishes button clicks to InputBox
import Button from 'react-bootstrap/Button';
import Col from 'react-bootstrap/Col';
import React, { useEffect } from 'react';
import Row from 'react-bootstrap/Row';

import { buttonEvent$ } from '../../shared/events/index';

const ButtonLayout = (props) => {
  
  const publishButtonEvent = (value, type) => {
    buttonEvent$.next({value, type});
  }

  useEffect(() => {
    let buttonHandlers = [];
    props.buttons.forEach(button => {
      let buttonEventHandler = (event) => {
        if (event.key === button.keyEvent) {
          buttonEvent$.next({value: button.value, type: button.type});
        }
      };
      document.addEventListener('keyup', buttonEventHandler);
      buttonHandlers.push(buttonEventHandler);
    });
    return () => {
      buttonHandlers.forEach(buttonHandler => {
        document.removeEventListener('keyup', buttonHandler);
      });
    }
  }, [props]);

  return (
    <div>
      <Row className='mt-2'>
        {props.buttons.map((button, index) => (
          <Col key={index} xs={2} className='my-1'>
            <Button className='w-100' variant={button.color} onClick={() => publishButtonEvent(button.value, button.type)}>
              {button.value.name}
            </Button>
          </Col>
        ))}
      </Row>
    </div>
  );
}

export { ButtonLayout };
