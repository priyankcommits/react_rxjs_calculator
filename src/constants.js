export const buttonTypes = {
    NUMBER: 'number',
    OPERATOR: 'operator',
    FUNCTION: 'function',
    RESET: 'reset',
}

export const operators = {
    PLUS: '+',
    MINUS: '-',
    MULTIPLY: 'x',
    DIVIDE: '÷',
}

const claculatorButtons = [
    {value: {name: 'sin', startsWith: 'sin(', endsWith: ')'}, type: buttonTypes.FUNCTION, color: 'secondary'},
    {value: {name: 'ln', startsWith: 'ln(', endsWith: ')'}, type: buttonTypes.FUNCTION, color: 'secondary'},
    {value: {name: '7'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: '8'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: '9'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: operators.DIVIDE}, type: buttonTypes.OPERATOR, color: 'secondary'},
    {value: {name: 'cos', startsWith: 'cos(', endsWith: ')'}, type: buttonTypes.FUNCTION, color: 'secondary'},
    {value: {name: 'log', startsWith: 'log(', endsWith: ')'}, type: buttonTypes.FUNCTION, color: 'secondary'},
    {value: {name: '4'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: '5'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: '6'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: operators.MULTIPLY}, type: buttonTypes.OPERATOR, color: 'secondary'},
    {value: {name: 'tan', startsWith: 'tan(', endsWith: ')'}, type: buttonTypes.FUNCTION, color: 'secondary'},
    {value: {name: 'π', startsWith: '', endsWith: 'π'}, type: buttonTypes.FUNCTION, color: 'secondary'},
    {value: {name: '1'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: '2'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: '3'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: operators.MINUS}, type: buttonTypes.OPERATOR, color: 'secondary'},
    {value: {name: 'pow', startsWith: '^', endsWith: ''}, type: buttonTypes.FUNCTION, color: 'secondary'},
    {value: {name: 'e', startsWith: '', endsWith: 'e'}, type: buttonTypes.FUNCTION, color: 'secondary'},
    {value: {name: '0'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: '.'}, type: buttonTypes.NUMBER, color: 'light'},
    {value: {name: 'AC'}, type: buttonTypes.RESET, color: 'primary'},
    {value: {name: operators.PLUS}, type: buttonTypes.OPERATOR, color: 'secondary'}
]

export default claculatorButtons;
