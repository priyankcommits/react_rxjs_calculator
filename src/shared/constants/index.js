// Contains all constants related to Appplication

const buttonTypes = {
  NUMBER: 'number',
  OPERATOR: 'operator',
  FUNCTION: 'function',
  CONSTANT: 'constant',
  RESET: 'AC',
  PARANTHESIS_OPEN: '(',
  PARANTHESIS_CLOSE: ')',
  POINT: '.',
  ANSWER: 'Ans',
  UNDO: 'Undo',
  REDO: 'Redo',
}

const operators = {
  PLUS: '+',
  MINUS: '-',
  MULTIPLY: 'x',
  DIVIDE: 'รท',
  POWER: 'pow',
  MODULUS: '%',
}

const calculatorButtons = [
  {value: {name: 'sin'}, type: buttonTypes.FUNCTION, color: 'secondary', keyEvent: 's'},
  {value: {name: 'ln'}, type: buttonTypes.FUNCTION, color: 'secondary', keyEvent: 'l'},
  {value: {name: '7'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '7'},
  {value: {name: '8'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '8'},
  {value: {name: '9'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '9'},
  {value: {name: operators.DIVIDE}, type: buttonTypes.OPERATOR, color: 'secondary', keyEvent: '/'},
  {value: {name: 'cos'}, type: buttonTypes.FUNCTION, color: 'secondary', keyEvent: 'c'},
  {value: {name: 'log'}, type: buttonTypes.FUNCTION, color: 'secondary', keyEvent: 'g'},
  {value: {name: '4'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '4'},
  {value: {name: '5'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '5'},
  {value: {name: '6'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '6'},
  {value: {name: operators.MULTIPLY}, type: buttonTypes.OPERATOR, color: 'secondary', keyEvent: '*'},
  {value: {name: 'tan'}, type: buttonTypes.FUNCTION, color: 'secondary', keyEvent: 't'},
  {value: {name: 'ฯ'}, type: buttonTypes.CONSTANT, color: 'secondary', keyEvent: 'p'},
  {value: {name: '1'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '1'},
  {value: {name: '2'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '2'},
  {value: {name: '3'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '3'},
  {value: {name: operators.MINUS}, type: buttonTypes.OPERATOR, color: 'secondary', keyEvent: '-'},
  {value: {name: operators.POWER}, type: buttonTypes.OPERATOR, color: 'secondary', keyEvent: '^'},
  {value: {name: 'e'}, type: buttonTypes.CONSTANT, color: 'secondary', keyEvent: 'e'},
  {value: {name: '0'}, type: buttonTypes.NUMBER, color: 'light', keyEvent: '0'},
  {value: {name: buttonTypes.POINT}, type: buttonTypes.POINT, color: 'light', keyEvent: '.'},
  {value: {name: buttonTypes.RESET}, type: buttonTypes.RESET, color: 'primary', keyEvent: 'r'},
  {value: {name: operators.PLUS}, type: buttonTypes.OPERATOR, color: 'secondary', keyEvent: '+'},
  {value: {name: buttonTypes.PARANTHESIS_OPEN}, type: buttonTypes.PARANTHESIS_OPEN, color: 'secondary', keyEvent: '('},
  {value: {name: buttonTypes.PARANTHESIS_CLOSE}, type: buttonTypes.PARANTHESIS_CLOSE, color: 'secondary', keyEvent: ')'},
  {value: {name: buttonTypes.UNDO}, type: buttonTypes.UNDO, color: 'warning', keyEvent: 'z'},
  {value: {name: buttonTypes.REDO}, type: buttonTypes.REDO, color: 'warning', keyEvent: 'x'},
  {value: {name: buttonTypes.ANSWER}, type: buttonTypes.CONSTANT, color: 'primary', keyEvent: 'a'},
  {value: {name: operators.MODULUS}, type: buttonTypes.OPERATOR, color: 'secondary', keyEvent: '%'},
]

export { buttonTypes, calculatorButtons, operators };
