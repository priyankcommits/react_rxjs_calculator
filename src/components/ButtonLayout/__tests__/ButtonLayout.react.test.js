// Snapshot testing
import React from 'react';
import renderer from 'react-test-renderer';

import { ButtonLayout } from '../ButtonLayout';
import { calculatorButtons } from '../../../shared/constants/index';

test('Button Layout Initial Rendering', () => {
  const component = renderer.create(
    <ButtonLayout buttons={calculatorButtons} />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});
