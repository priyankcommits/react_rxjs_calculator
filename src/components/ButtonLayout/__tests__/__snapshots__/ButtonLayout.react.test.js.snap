// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Layout Initial Rendering 1`] = `
<div>
  <div
    className="mt-2 row"
  >
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        sin
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        ln
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        7
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        8
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        9
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        cos
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        log
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        4
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        5
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        6
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        x
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        tan
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        ฯ
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        1
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        2
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        3
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        -
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        pow
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        e
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        0
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-light"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        .
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        +
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        (
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        )
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-warning"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Undo
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-warning"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Redo
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Ans
      </button>
    </div>
    <div
      className="my-1 col-2"
    >
      <button
        className="w-100 btn btn-secondary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        %
      </button>
    </div>
  </div>
</div>
`;
